### DISCUSSION TITLE
자료구조 - 트리 - 서희&재찬

### DISCUSSION THREAD

**Discussion by kjc6735** [Shortudy-10th/tech-for-developer#62] [질의 응답]
- B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요.
- 레드-블랙트리에 대해 설명해주세요.
- AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.
- 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요
- BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.
- 해당 트리의 높이와 차수를 말해주세요.
![tree](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/8e295580-04d5-47f9-b5c0-1b886cfacb5d)

**Comment by kjc6735** [comment]
### B트리와 B+트리의 구조와 그 차이점에 대해 설명해주세요.

<br />

<blockquote>
B-트리
구조 : B-트리는 내부 노드와 리프 노드가 모두 데이터를 가지고 있다.
내부 노드는 데이터를 가지며, 이 데이터를 통해 트리의 구조를 관리

B+-트리
B+-트리는 내부 노드가 데이터를 가지지 않고, 모든 데이터는 리프 노드에만 저장
내부 노드는 데이터를 참조하기 위한 역할

차이점 
데이터 저장 - 비트리는 리프 노드, 브랜치 노드 모두 데이터 저장 가능하지만, b+트리는 오직 리프 노드만 저장 가능

풀 스캔 시, B+tree는 리프 노드에 데이터가 모두 있기 때문에 한 번의 선형탐색만 하면 되기 때문에 B-tree에 비해 빠르다. B-tree의 경우에는 모든 노드를 확인해야 한다. 

키 중복 - 비트리의 경우 중복 키가 없지만, b+트리의 경우에는 있을 수 있다.

비트리는 리프 노드끼리의 연결 리스트가 없다.
b+트리는 리프 노드끼리 연결 리스트 
</blockquote>

참고
https://seunggi92.tistory.com/39
https://zorba91.tistory.com/293

**Comment by kjc6735** [comment]
### 레드-블랙트리에 대해 설명해주세요.

<blockquote>
  레드-블랙 트리의 특징은 삽입, 삭제 동안 트리의 모양이 균형 잡히도록 각 노드들이 red나 black색상을 가진다는 것이다. 검색, 삽입, 삭제 시 최악의 경우에서도 모두 O(logn)이 보장되는 자료 구조이다.
  
  특징
 루트 노드와 NIL노드는 항상 블랙이다.
 레드 노드의 자식은 항상 블랙이며, 블랙 노드의 자식은 아무거나 상관없다.
 루트 노드에서 모든 리프 노드까지의 경로에 대해서 거치는 블랙 노드의 숫자는 같다.


삽입 과정 
삼촌 노드(부모의 부모노드의 다른 자식 노드)의 색에 따라 좌우된다.
삼촌 노드가 검정이면 restructing, 빨간 색이면 recoloring이 사용된다.

리스트럭팅 과정
1) 삽입된 노드와 그 부모, 그 조부모 노드를 오름차순으로 정렬합니다.

2) 가운데 값을 부모로 만든 뒤, 나머지 둘을 자식으로 만들어줍니다.

3) 이때 부모는 검은색 노드, 두 자식들은 빨간색 노드로 만듭니다.

시간 복잡도 O(logN)

리컬러링 과정 
1) 현재 삽입된 노드의 부모와 삼촌을 검은색으로 만든 후, 조부모를 빨간색으로 만들어줍니다.

2) 만약 조부모가 루트 노드가 아니라면, 조부모의 부모가 빨간색일 수 있고, 이는 다시 Double Red 상황을 일으킬 수 있어 다시 한번 위로 돌아가 두 가지 스킬에 대해 고민해봐야 합니다. (propagation)



삼촌 노드가 없을 경우에는 NIL노드가 자식 노드이고, NIL노드 또한 검정 노드이기 때문에 Restructing을 수행하면 된다.
</blockquote>

참고

https://suhwanc.tistory.com/197

**Comment by kjc6735** [comment]
### AVL트리에 대해 해당 트리의 삽입, 검색, 삭제의 시간 복잡도를 포함하여 설명해주세요.

<blockquote>
  AVL 트리는 높이가 logN으로 유지되기 때문에 삽입, 검색, 삭제 시간이 O(logN)이다.
삽입 시 루트까지 올라가면서 Balance Factor가 -2, 2가 되는 경우 로테이트를 진행한다. 회전의 시간 복잡도는 O(1)이고, 삽입 동작이 O(logN)이다. 
AVL트리는 항상 균형잡혀 있기 때문에 높이는 항상 O(logN)이고, 검색 연산의 평균, 최악의 시간 복잡도는 O(logN)이 된다.
삭제도 마찬가지로 삭제 후 균형 맞춰주는 회전은 O(1) 이며, 삭제 노드 찾는 것이 O(logN)이기 때문에 시간 복잡도는 O(logN)이다.
</blockquote>

**Comment by LireEruel** [comment]
### 트리 순회를 위해 사용하는 방법에는 무엇이 있는지 설명해주세요

**1. 전위 순회 (Preorder Traversal):**
- 루트(root)를 먼저 방문합니다.
- 왼쪽 서브트리를 순회합니다.
- 오른쪽 서브트리를 순회합니다.
- 순회 과정: 루트 -> 왼쪽 -> 오른쪽

**2. 중위 순회 (Inorder Traversal):**
- 왼쪽 서브트리를 먼저 순회합니다.
- 루트를 방문합니다.
- 오른쪽 서브트리를 순회합니다.
- 순회 과정: 왼쪽 -> 루트 -> 오른쪽
- 이진 검색 트리에서 중위 순회는 정렬된 순서로 노드를 반환합니다.

**3. 후위 순회 (Postorder Traversal):**
- 3.왼쪽 서브트리를 먼저 순회합니다.
- 오른쪽 서브트리를 순회합니다.
- 루트를 방문합니다.
- 순회 과정: 왼쪽 -> 오른쪽 -> 루트

**4.레벨 순회 (Level Order Traversal):**
- 트리의 레벨별로 노드를 방문합니다.
- 보통 큐(queue)를 사용하여 트리의 각 레벨에서 노드를 방문하면서 그 자식 노드를 큐에 추가합니다.
- 이 방법은 너비 우선 탐색(BFS)과 관련이 있습니다.

**Comment by LireEruel** [comment]
### BST(Binary Search Tree)와 Binary Tree에 대해 설명해주세요.

**Reply by LireEruel** [reply]
**Binary Tree (이진 트리)**
이진 트리는 각 노드가 최대 두 개의 자식 노드를 가지는 트리 구조입니다. 이진 트리의 노드는 보통 루트, 왼쪽 자식, 오른쪽 자식으로 구성됩니다.

이진 트리의 특징:

- 각 노드는 최대 두 개의 자식 노드를 가질 수 있음.
- 왼쪽과 오른쪽 자식의 순서가 중요함.
- 이진 트리는 균형 트리일 수도 있고, 그렇지 않을 수도 있음.
- 순회 방법으로는 전위, 중위, 후위 순회 등이 있음.

**Binary Search Tree (BST, 이진 검색 트리)**
이진 검색 트리는 이진 트리의 한 형태로, 특정한 속성을 가진다는 점에서 다릅니다.

BST의 특징:

- 각 노드의 왼쪽 서브트리에 있는 모든 노드의 키는 노드의 키보다 작아야 함.
- 각 노드의 오른쪽 서브트리에 있는 모든 노드의 키는 노드의 키보다 커야 함.
- 왼쪽과 오른쪽 서브트리 모두 이진 검색 트리여야 함.
- 중복된 노드가 없어야 함.

**Reply by LireEruel** [reply]
### BST 장단점에 대해 알려주세요.

**BST의 장점:**

- 키의 탐색, 삽입, 삭제 등의 연산이 평균적으로 O(logn) 시간에 수행될 수 있음. 단, 트리가 균형 잡힌 경우에만 해당.
- 중위 순회를 사용하면 키 값을 정렬된 순서로 얻을 수 있음.

**BST의 단점:**

- 트리가 한쪽으로 치우쳐진 경우(예: 트리가 연결 리스트처럼 되어 있는 경우), 연산의 시간 복잡도가 O(n)이 될 수 있음.

**Comment by LireEruel** [comment]
### 해당 트리의 높이와 차수를 말해주세요.
![tree](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/8e295580-04d5-47f9-b5c0-1b886cfacb5d)

**Comment by LireEruel** [comment]
트리의 높이는 루트에서 가장 깊은 노드까지의 길이입니다.
따라서 이 트리의 높이는 4 입니다.

트리의 차수는 트리의 최대 차수를 말합니다.
따라서 이 트리는 3 입니다.
